src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelMetrics.cs

```diff
@@ -0,0 +1,240 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.Extensions.Metrics;
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
+
+internal sealed class KestrelMetrics
+{
+    // Note: Dot separated instead of dash.
+    public const string MeterName = "Microsoft.AspNetCore.Server.Kestrel";
+
+    private readonly Meter _meter;
+    private readonly UpDownCounter<long> _currentConnectionsCounter;
+    private readonly Histogram<double> _connectionDuration;
+    private readonly Counter<long> _rejectedConnectionsCounter;
+    private readonly UpDownCounter<long> _queuedConnectionsCounter;
+    private readonly UpDownCounter<long> _queuedRequestsCounter;
+    private readonly UpDownCounter<long> _currentUpgradedRequestsCounter;
+    private readonly Histogram<double> _tlsHandshakeDuration;
+    private readonly UpDownCounter<long> _currentTlsHandshakesCounter;
+
+    public KestrelMetrics(IMeterFactory meterFactory)
+    {
+        _meter = meterFactory.CreateMeter(MeterName);
+
+        _currentConnectionsCounter = _meter.CreateUpDownCounter<long>(
+           "current-connections",
```

Comment:
Is there a pattern which the .NET team would want to encourage folks to use? How do we plan to organize these names? @noahfalk FYI

FYI OpenTelemetry API is quite flexible https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument-name-syntax.

Reply:
I've followed the naming standard that .NET has for event counters.

Meter name: Dot separated (like a .NET namespace or project)
Counter name: Dash separated
Tag name: Dash separated

We'll put these names through our API review process. You might be interested in taking a look: https://github.com/dotnet/aspnetcore/issues/47536

Reply:
> We'll put these names through our API review process. You might be interested in taking a look: #47536

Nice! Thanks!

Reply:
FYI @JamesNK I don't have permission/access to resolve the comment, feel free to mark it as resolved.

![image](https://user-images.githubusercontent.com/17327289/230098021-83e326d1-e08f-41b0-a17e-ab82fddbe1f5.png)